### PR TITLE
[WIP] firmware_retraction: Implementation of tuneable G10/G11 firmware retraction

### PR DESCRIPTION
**Pull request by Lenbok**
Support getting and setting of retraction parameters without having to restart klippy. Two main benefits compared to using `gcode_macro`:
* Retraction settings can be adjusted during a print (primarily useful during calibration)
* Retraction settings can be set in the per-filament gcode section of your slicer, allowing different settings for different types of plastics.

Signed-off-by: Len Trigg <lenbok@gmail.com>

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  FYI, there's been a lot of interest recently in the use of the Jinja2 template system for similar capabilities.  (It doesn't have the features done here, but the general ability to set and query a "variable" has been identified as being useful in several modules.)  If we added generic "template variables", would that be sufficient for your needs?

-Kevin

**Comment by Lenbok** [comment]
If the question is "would this feature be implementable in jinja?", the answer is probably yes (but I can't really say, if the capability isn't there yet). If we later decide to extend this so that the retraction settings are to be per-extruder for better multi-material support, then I suspect it may not be possible.

If the question is "should this feature be in the main klipper repo (in python) vs something the user is expected to find themselves (and add to their config via jinja)?", I think that for relatively core features (by which I mean other firmwares have this capability and it is directly related to features that slicers support), they should be available via the main klipper repo rather than something each user has to cook up for themselves.

If the question is "should this feature be implemented via jinja templating instead of python?", I think that  it should be easy to extend clipper by a mechanism that the user is comfortable with, and in this regard, python is a language that a large fraction of the programming population knows and likes, whereas jinja is one that barely anyone knows.

Getting off-topic, what does jinja bring to the table? Is it easier to test? Is it easier to debug? Is it easier to share code?

**Comment by mcmatrix** [comment]
If you'll add `get_status` method to your module.
Something like this
```py
def get_status(self, eventtime):
    return {
        "retract_length": self.retract_length,
        "retract_speed": self.retract_speed,
        "unretract_extra_length": self.unretract_extra_length,
        "unretract_speed": self.unretract_speed
    }
```
Then it will be easy later add a menu for firmware retraction where you can adjust these params.

**Comment by BlackStump** [comment]
is it possible to include Z hop or the option for Z, it is a feature I use on Delta printers

**Comment by nythil** [comment]
I think it's wrong to blindly assume the usage of relative/absolute coordinates ("G91\nG1 E%d F%d\nG90"). Imagine that one already was in relative mode (G91) and wished to stay in it after retraction/unretraction. This code would force absolute move and break the g-code that follows. You could use toolhead.move() method instead.

**Comment by KevinOConnor** [comment]
Thanks for the feedback.  I have uploaded PR #1689 with documentation and enhancements to the Jinja2 template system.

> If the question is "should this feature be in the main klipper repo (in python) vs something the user is expected to find themselves (and add to their config via jinja)?"

That's a great question.  In this case, it looks to me like the G10/G11 commands seem to more closely fit user defined macros.  That is, they are actions implemented via g-code commands that different users will likely want to implement in slightly different ways.

In general, I prefer for code to be implemented in Python.  However, if the code is only relaying configuration to g-code commands then I fear that the resulting Python code could become an endless list of possible user configurations.  A macro may ultimately be more flexible in that case.

FWIW, here's what an equivalent macro might look like after PR #1689:
```
[gcode_macro G10]
variable_retract_length: 10
gcode:
  SAVE_GCODE_STATE
  G91
  G1 E-{retract_length} F60
  RESTORE_GCODE_STATE

[gcode_macro G11]
gcode:
  SAVE_GCODE_STATE
  G91
  G1 E{printer["gcode_macro G10].retract_length} F60
  RESTORE_GCODE_STATE
```

-Kevin

**Comment by BlackStump** [comment]
> I think it's wrong to blindly assume the usage of relative/absolute coordinates ("G91\nG1 E%d F%d\nG90"). Imagine that one already was in relative mode (G91) and wished to stay in it after retraction/unretraction. This code would force absolute move and break the g-code that follows. You could use toolhead.move() method instead.

so as a alternative could this be used
 ("SAVE_GCODE_STATE\G91\nG1 E%d F%d\RESTORE_GCODE_STATE")
I apologize I am not a programmer at all just trying to fudge things together.

**Comment by Lenbok** [comment]
@BlackStump It'll be easy to add the Z lift at the same time. I saw that is in the firmware retraction implementation of other firmware but have never used that option myself.

@KevinOConnor That save/restore gcode state definitely seems the way to go, I'll update this PR once that one has merged. The macro example you posted doesn't prevent multiple successive retractions or successive unretracts, so perhaps that's easier via python. Once Z-lift is added, that should be basically parity with the other firmware implementations, and doubt that much further user-tinkering is expected.

I'm curious if you have any vision of how python and/or macro extensions can be more easily shared between klipper users it they are not present in the main klipper repo. The advantage of being somewhat permissive in what is included in the main repo is that it's easily searchable and there is more confidence that they work with the current klipper than something cribbed off the net. But things could get unmanageable over time. I do think that for platforms of any complexity/power some kind of plugin repository system really helps things to take off (I only wish OpenSCAD had such a system).

**Comment by BlackStump** [comment]
@Lenbok 
This is what I put together and tested to work fine
https://github.com/BlackStump/klipper/blob/master/klippy/extras/firmware_retraction.py

**Comment by Lenbok** [comment]
Incorporated those suggestions. Since it uses the gcode state, it'll need #1684 to be merged first.

**Comment by KevinOConnor** [comment]
Thanks.

> The macro example you posted doesn't prevent multiple successive retractions or successive unretracts, so perhaps that's easier via python.

I guess I don't really understand what G10/G11 is used for.  That said, I don't have any objections to merging this.  I think it would be good if the following items were addressed though:
- It'd be preferable to add something like `NAME=_retract_state` to the SAVE_GCODE_STATE command (otherwise it may confuse users that have used SAVE_GCODE_STATE for some other purpose in a calling macro).
- The config/example-extras.cfg file should be updated with a description of the config parameters.
- The docs/G-Codes.md file should be updated to describe the new commands.

> I'm curious if you have any vision of how python and/or macro extensions can be more easily shared between klipper users it they are not present in the main klipper repo. 

Another excellent question.  I don't think a full blown plugin system will work well right now - the "extras" modules currently have too much power to alter the state of Klipper so it would be extremely difficult for us to diagnose issues if a user is using an external module.  At some point it might make sense to support that, but I don't think that point is in the near future.  The best advice I have is to merge into the main repo (either in klippy/extras/ or in config/sample-macros.cfg).  I understand that isn't always a great solution though.

-Kevin

**Comment by BlackStump** [comment]
The thing I like about using firmware retraction apart from on the fly adjustable is I can have a common retract setting across all slicers that support firmware retraction and most do have support.
Simplified3D is a little annoying it uses M103/M101 instead of G10/G11 so it would be nice to add M103/M101 as aliases for G10/G11 
If I want to try a different slicer I tick firmware retraction , add my start and end macros and I can then focus on the attributes of the slicer.

**Comment by Lenbok** [comment]
@BlackStump I've been testing the z_hop feature on my delta and it (in conjunction with the way the GCODE_STATE command works) is causing klipper to abort the print on the first layer. It looks to be on the G10/G11 associated with the first layer when the z_hop value is larger than the first layer height. Have you tried that scenario?

On the first G10, we do a SAVE_GCODE_STATE, then execute the retraction command (which lifts the Z axis by z_hop), then RESTORE_GCODE_STATE (which resets klipper's notion of the XYZ coords to where they were prior to the start of the G10, i.e. it now thinks the Z is lower than it actually is, by z_hop).

Then when we encounter the first G11, we do the SAVE_GCODE_STATE, then execute the unretraction command (which commands the Z back down by the z_hop amount). Now since klipper thinks the Z is lower than it actually is, it aborts thinking that the Z move would drive the head through the bed.

@KevinOConnor Is there a way to make the RESTORE_GCODE_STATE not restore the XYZ position? In our use case here, we are only using it so we can preserve the G90/G91 coordinate mode (and probably also move speeds)

**Comment by BlackStump** [comment]
that is interesting I have only tested on the ender3 not the delta as yet but I have a first layer height of 0.3 and following layer height of 0.2 and z_hop of 0.2 and not had a issue I have on the fly set z_hop to 4mm to visually see the z_hop and did not experience a problem. I will test the scenario you have described on the ender3 to see if I experience the same

**Comment by Lenbok** [comment]
@BlackStump the z_hop value was previously being printed as integer (i.e. 0.2 would have had the same effect as 0), so you would have to have been using a value higher than 1.0 to have seen it (I was testing with a high value I could visually see the lift).

**Comment by BlackStump** [comment]
odd that I can not reproduce the error 1.4mm z_hop still no problems

**Comment by Lenbok** [comment]
It will also depend on what your configuration `position_min` is set to. If that is permissive enough then klipper won't issue the complaint.

This Z hop needs some adjustment (hence the change to [WIP]). Simply having the G10 and G11 do relative Z movements is not correct, since if you have a sequence like:
```
G10
G1 Z0.250 F24000.000
G1 X20.000 Y-25.696 F24000.000
G11
```
If the G10 does a relative Z lift, then the intervening command does an absolute Z command, we can't have the G11 issue a relative Z drop, since that will leave us at 0.25-z_hop. It is probably better to have G10 do a `SET_GCODE_OFFSET Z_ADJUST=<z_hop>` and G11 to reverse this with `SET_GCODE_OFFSET Z_ADJUST=-<z_hop>`

**Comment by BlackStump** [comment]
I agree that does seem like a better idea

**Comment by Lenbok** [comment]
After setting my `minimum_z_position` (which I believe is the delta equivalent of cartesian stepper_z `position_min`) to be more tolerant, the prints do work fine for me also when using z_hop, using the code in the PR as it is (since the current gcode state position save/restore sequence is effectively enacting a z offset). I'm not sure how the actual SET_GCODE_OFFSET interacts with the minimum positions, will need to try it to see whether it's any better.

**Comment by CapnBry** [comment]
I look forward to seeing this integrated once it and its dependencies are complete since it makes maintaining my Slic3r profiles a lot easier if I don't have to have separate printer profiles for every filament that needs different retraction settings. However, I feel the precision of the z-hop value in the injected gcode is insufficient in a9554cc. I z-hop either half a layer height or a full layer height which is 0.12mm or 0.24mm respectively. The 0.1f precision chosen is somewhat arbitrary and probably should follow convention as to what slicers put in gcode, which is 0.3f for Z. Obviously that level of precision probably won't ever be needed for a z-hop, but parsing two extra characters in the gcode to allow it to never be a problem has no real performance impact.

I'd also recommend 0.5f for E if you're making a change, just to match other gcode output but I don't think that would ever make a difference.

File: klippy/extras/firmware_retraction.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            "unretract_extra_length": self.unretract_extra_length,
+            "unretract_speed": self.unretract_speed,
+            "z_hop": self.z_hop
+        }
+
+    def cmd_SET_RETRACTION(self, params):
+        self.retract_length = self.gcode.get_float(
+            'RETRACT_LENGTH',
+            params, self.retract_length, minval=0.)
+        self.retract_speed = self.gcode.get_int(
```
Comment (KevinOConnor) [inline klippy/extras/firmware_retraction.py]: Use get_float() here.  Also, please use values in mm/s (not mm/min - g-code is just nuts).

File: klippy/extras/firmware_retraction.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            'UNRETRACT_EXTRA_LENGTH',
+            params, self.unretract_extra_length, minval=0.)
+        self.unretract_speed = self.gcode.get_int(
+            'UNRETRACT_SPEED',
+            params, self.unretract_speed, minval=1)
+        self.z_hop = self.gcode.get_float(
+            'Z_HOP',
+            params, self.z_hop, minval=0.)
+        self.unretract_length = self.retract_length
+        + self.unretract_extra_length
```
Comment (KevinOConnor) [inline klippy/extras/firmware_retraction.py]: Similar multi-line corruption.

File: klippy/extras/firmware_retraction.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        self.is_retracted = False
+        #self.cmd_GET_RETRACTION(params)
```
Comment (KevinOConnor) [inline klippy/extras/firmware_retraction.py]: Dead code?

File: klippy/extras/firmware_retraction.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                  self.unretract_extra_length, self.unretract_speed,
+                  self.z_hop))
+        self.gcode.respond_info(msg)
+
+    def cmd_G10(self, params):
+        if not self.is_retracted:
+            self.gcode.run_script_from_command(
+                "SAVE_GCODE_STATE NAME=_retract_state\n"
+                "G91\n"
+                "G1 E-%.1f F%d Z%.1f\n"
```
Comment (KevinOConnor) [inline klippy/extras/firmware_retraction.py]: As mentioned elsewhere, probably better to use more precision in the floats.

**Comment by KevinOConnor** [comment]
> Is there a way to make the RESTORE_GCODE_STATE not restore the XYZ position?

The RESTORE_GCODE_STATE only moves the toolhead back if the MOVE=1 parameter is specified.  So, it doesn't restore XYZ position by default.  If you've run into an issue, can you reproduce on pristine code and post the Klipper log from the session?

-Kevin

**Comment by BlackStump** [comment]
@Lenbok 
@KevinOConnor 
This is the code I am currently using,
https://github.com/BlackStump/klipper/blob/master/klippy/extras/firmware_retraction.py
 I am not getting any errors apart from,
Move exceeds maximum extrusion (30.066mm^2 vs 8.000mm^2)
as you will see by the log I have 2.5 mm retraction
It retracts fine but unretract it errors, 
If I change my max_extrude_cross_section=40
then of course no errors, is this expected behavior with retraction?
I have been using max_extrude_cross_section=8 prior to klipper updates and firmware retraction
[klippy.log](https://github.com/KevinOConnor/klipper/files/3255035/klippy.log)

**Comment by KevinOConnor** [comment]
@BlackStump - you'll need to issue the z hop in a G1 separate from the retract/unretract.  Gcode is terrible - if you issue an E movement along with an XYZ movement then the F is the speed of the toolhead move and the extruder must sync to that speed.  The errors are the result of that confusion.

-Kevin

**Comment by BlackStump** [comment]
@KevinOConnor 
That makes perfect sense and yes Gcode is terrible, I will make some changes
Thanks

**Comment by BlackStump** [comment]
@Lenbok 
@KevinOConnor 
Tested to work back at max_extruder_cross_section=8
https://github.com/BlackStump/klipper/blob/master/klippy/extras/firmware_retraction.py

**Comment by Lenbok** [comment]
Updated based on above comments/fixes (I haven't tested myself yet)

**Comment by FHeilmann** [comment]
I think the G92 E0 should be removed here. Users that use absolute extrusion instead of relative will run into errors here if the retract and unretract suddenly zeroes their extruder. Switching the extruder to relative, retracting and then restoring the previous state should be sufficient here.

**Comment by KevinOConnor** [comment]
Thanks!

I agree the `G92 E0` should be removed, but it is harmless.  The `RESTORE_GCODE_STATE` command will restore the previous g-code "origin".  Also, FYI, M83 isn't needed as G91 puts the extruder into relative mode.

-Kevin

**Comment by KevinOConnor** [comment]
Oh - as an additional note - I suspect users may want a z_hop speed.  The extruder speed of retraction is unlikely to match the desired z move speed.

-Kevin

**Comment by FHeilmann** [comment]
> FYI, M83 isn't needed as G91 puts the extruder into relative mode.

I didn't know that. Is that according to gcode spec or just something that klipper does?

**Comment by KevinOConnor** [comment]
Hah, "gcode spec", very funny!

The last time I looked, it was the behaviour of other common firmware that slicers, OctoPrint, and similar depended on.

-Kevin

**Comment by BlackStump** [comment]
I have had a go at adding the z_hop speed addition, I have not tested as my printer is well printing.
A question in the GET_RETRACTION section is it necessary to have that level of precision for the message?
There is precision in the adjustment. 
https://github.com/BlackStump/klipper/blob/master/klippy/extras/firmware_retraction.py

**Comment by KevinOConnor** [comment]
I've received reports that z_hop does not work correctly in the firmware_retraction module.  Specifically, it appears that if the slicer performs a layer change after a G10, that layer change is not realized on the following G11.  So, z_hop raises the head, the layer change also raises the head (but not the full amount), and then G11 lowers the head (partially into the print).

We'll need to either fix this behaviour or remove the z_hop support.

-Kevin

**Comment by BlackStump** [comment]
@KevinOConnor 
I have not noticed that behavior, I am using retraction with z-hop and with the z-hop speed addition
I have done a lot of printing since it was included.
how could I specifically test the reported behavior?

**Comment by KevinOConnor** [comment]
Unfortunately, I don't have the exact details.  I believe the following shows the problem though:
```
g28
g90
g1 x0 y0 z20
g10
g1 x10 y10 z30
g11
```

At the end of the above sequence, I believe the toolhead should be at Z=30.  However, it would instead be at position Z=(30-z_hop).

-Kevin

**Comment by BlackStump** [comment]
I look at that and think why and when would a slicer issue a retract, unretract in that situation? I think someone is using a G10,G11 command for a unintended use?

G10,M103, G11,M101 are commands marked by a slicer for print retractions instead of using the a slicers 
software retraction however the slicer still issues the G10, G11 in the correct locations. just as it would if it was issuing software retracts.
I guess I am saying if the person used a slicers retraction, a retract and unretract would not be inserted there?
Gcode if you manually issue the wrong command in the wrong location you will have unintended consequences?

**Comment by Maglin** [comment]
I recently tried to use the FW retraction and saw an issue where it seemed the first retraction didn’t happen directly after my start print macro.  And the following G11 did happen.  I was using 4.0mm retraction @ 40mms and 1mm zhop at 50mms.  I didn’t have the time to test what was the full cause.  I’ll look at it this coming weekend if this hasn’t been resolved by then.

**Comment by BlackStump** [comment]
I would be interested to see the offending gcode and what slicer produced it, if manually inserting a firmware retraction then location of that insertion is as important as any other gcode.
If I read the code a G11 will not be acted on unless a previous G10 has been issued?
I have not noticed the behavior, but I let the slicer issue the commands.
I only use slicers that have the option for firmware retraction but also are not using multi extruders. If there is a problem I will have to hand it back to programmers as I am out of my depth now.

Thanks

**Comment by BlackStump** [comment]
@Lenbok 
@KevinOConnor 
the z_hop speed has not been added as yet.

**Comment by CapnBry** [comment]
@BlackStump If it helps at all, I can say for sure that Slic3r PE / PrusaSlicer generates the offending sequence if `Printer Settings -> Extruder 1 -> Retract on layer change` is enabled:
```
G1 X111.087 Y134.815 E87.04742
; stop printing object hm-case-43.stl id:0 copy 0
M106 S153
; layer 2
G10 ; retract
G92 E0
G1 Z0.720 F15000.000
@Object hm-case-43.stl id:0 copy 0
G1 X116.479 Y141.906
G11 ; unretract
G92 E0
G1 F3360
G1 X117.363 Y141.165 E0.06655
```
If the option is turned off, it still generates the retract when changing layers (and it needs to travel) but instead it issues it this way
```
G1 X111.087 Y134.815 E87.04742
; stop printing object hm-case-43.stl id:0 copy 0
M106 S153
; layer 2
G1 Z0.720 F15000.000
@Object hm-case-43.stl id:0 copy 0
G10 ; retract
G92 E0
G1 X116.479 Y141.906
G11 ; unretract
G92 E0
G1 F3360
G1 X117.363 Y141.165 E0.06655
```

I can confirm that with the first gcode with a 0.24mm layer height and a 0.24mm z_hop, the z height after the layer change is certainly not Z0.720 because the extruder starts skipping steps and jamming immediately, indicating it did not actually change layers.

**Comment by Arksine** [comment]
I can provide two examples of how firmware_retraction's zhop can negatively affect your print.  The first is with Cura.  After your start gcode executes, Cura will generate gcode something like the following:
```
G10
G0 X50 Y50 Z.2  ;move tool to beginning of first move
G11
```
Because Cura moves the tool to an absolute Z position prior to the detraction, zhop will move relative down the zhop amount.  If you zhop is greater than your initial layer height this will cause a nozzle crash.

The second example affects every slicer I use.  When a print finishes, the slicer generates a retraction.  If you start a subsequent print, the initial G10 is ignored because it thinks its already in a retracted state.  Howeve the G11 executes, which will zhop down and once again result in a nozzle crash.

I think firmware_retraction needs to have a check to make sure it doesn't zhop back down if the absolute z position changes while the tool is retracted.  An exception to this might be if the gcode offset was applied during the retracted state, in that case you would still want to zhop back down.

**Comment by BlackStump** [comment]
will this help the first part?
def cmd_G10(self, params):
        if not self.is_retracted:
            self.gcode.run_script_from_command(
                "SAVE_GCODE_STATE NAME=_retract_state\n"
                "G92 E0\n"
                "G91\n"
                "G1 E-%.5f F%d\n"
                "RESTORE_GCODE_STATE NAME=_retract_state\n"
                "SET_GCODE_OFFSET Z_ADJUST=%0.5f MOVE=1 MOVE_SPEED=%d"
                % (self.retract_length, self.retract_speed*60, self.z_hop,
                   self.z_hop_speed))
            self.is_retracted = True

    def cmd_G11(self, params):
        if self.is_retracted:
            self.gcode.run_script_from_command(
                "SAVE_GCODE_STATE NAME=_retract_state\n"
                "G92 E0\n"
                "G91\n"
                "G1 E%.5f F%d\n"
                "RESTORE_GCODE_STATE NAME=_retract_state\n"
                "SET_GCODE_OFFSET Z_ADJUST=-%0.5f MOVE=1 MOVE_SPEED=%d"
                % (self.unretract_length, self.unretract_speed*60, self.z_hop,
                   self.z_hop_speed))

**Comment by KevinOConnor** [comment]
FYI, I removed support for z_hop (commit 0343d0c7) until we can get this sorted out.

-Kevin

**Pull request merged**